File: crates/klyron_postgres/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::PostgresBuilder::Impl {
public:
  Klyron::PostgresConfig config;
  bool verbose = false;
};

Klyron::PostgresBuilder::Klyron::PostgresBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::PostgresBuilder& Klyron::PostgresBuilder::withConfig(const Klyron::PostgresConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::PostgresBuilder& Klyron::PostgresBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::PostgresBuilder::Klyron::PostgresInstance Klyron::PostgresBuilder::build() {
  Klyron::PostgresInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_postgres/bindings/rust/builder.rs
//! Builder pattern for klyron_postgres

use crate::types::Klyron::PostgresConfig;

#[derive(Debug, Default)]
pub struct Klyron::PostgresBuilder {
    config: Option<Klyron::PostgresConfig>,
    verbose: bool,
}

impl Klyron::PostgresBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::PostgresConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::PostgresInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::PostgresInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::PostgresInstance {
    pub config: Klyron::PostgresConfig,
    pub verbose: bool,
}
